[FLINK-31776][runtime] Introduces LeaderElection interface

XComp opened the pull request:
PR order:
- https://github.com/apache/flink/pull/21742
- https://github.com/apache/flink/pull/22379
- https://github.com/apache/flink/pull/22422
- https://github.com/apache/flink/pull/22380
- https://github.com/apache/flink/pull/22623
- === this PR === FLINK-31776
- https://github.com/apache/flink/pull/22390
- https://github.com/apache/flink/pull/22404
- https://github.com/apache/flink/pull/22601
- https://github.com/apache/flink/pull/22642
* https://github.com/apache/flink/pull/22640
* https://github.com/apache/flink/pull/22656
* https://github.com/apache/flink/pull/22661

## What is the purpose of the change

This change introduces a new interface `LeaderElectionService.LeaderElection` which serves as a proxy between the `LeaderContender` and the `LeaderElectionService`. Future PRs will introduce a contender ID which can hidden in `LeaderElection`. That way, `LeaderContender` wouldn't be aware of this implementation detail of the `DefaultLeaderElectionService`

## Brief change log

* Moved `confirmLeadership` and `hasLeadership(UUID)` into new interface
* `LeaderContender` registration moved out of `LeaderElectionService.createLeaderElection(..)` into `LeaderElection.startLeaderElection`
* `LeaderElection` implements `AutoCloseable` to cover the deregistration of the `LeaderContender` from the `LeaderElectionService`
* Introduced `AbstractLeaderElectionService` as a generic implementation of `LeaderElectionService` that deals with the `LeaderElection` lifecycle management and provdes helper methods

## Verifying this change

* Tests were updated and should still pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionService.java:
```diff
@@ -53,24 +50,35 @@ public interface LeaderElectionService {
+     * LeaderContender}.
      */
-    void confirmLeadership(UUID leaderSessionID, String leaderAddress);
+    interface LeaderElection {
```
Working on the follow-up PRs (more specifically: https://github.com/apache/flink/pull/22404)  properly (especially refactoring all the tests), I start to believe that having this as a sub-interface isn't the best choice. `StandaloneLeaderElectionService` will be replaced by a `StandaloneLeaderElection` implementation. The test cases which rely on a `LeaderElection` do not necessarily rely on the `LeaderElectionService`, either. Having them separate and not that closely-coupled feels like a more reasonable thing to do. :thinking: WDYT?

zentol replied:
Cant hurt to separate them :shrug:

XComp commented:
@flinkbot run azure

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/util/DocumentingDispatcherRestEndpoint.java:
```diff
@@ -86,20 +86,20 @@ public List<Tuple2<RestHandlerSpecification, ChannelInboundHandler>> initializeH
 
-    private enum NoOpElectionService implements LeaderElectionService {
-        INSTANCE;
+    private static class NoOpElectionService extends AbstractLeaderElectionService {
```
Yes, we can do that. :+1: The `LeaderElectionService` implementation will go away with PR #22404 anyway.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalTest.java:
```diff
@@ -120,6 +121,7 @@ void testConnectingAddressRetrievalWithDelayedLeaderElection() throws Exception
         long sleepingTime = 1000;
 
         LeaderElectionService leaderElectionService = null;
+        LeaderElection leaderElection = null;
```
You're right - it's not used for now but will be used with #22390 where we move the deregistration of the `LeaderElection` from `LeaderElectionService.stop()` into `LeaderElection.close()`. I added some documentation to reflect that.

zentol replied:
but why add the field now instead of in the other PR? :thinking:

XComp replied:
I haven't pushed the change related to [the comment above](https://github.com/apache/flink/pull/22384#discussion_r1200287477), yet. I'm working on it. The local variable will be removed.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingContender.java:
```diff
@@ -41,13 +42,20 @@ public TestingContender(
     }
 
+    public LeaderElection startLeaderElection() throws Exception {
+        leaderElection = leaderElectionService.createLeaderElection();
```
`LeaderElection.startLeaderElection` needs to be called outside of the contender's constructor for now. The reason for this is that we have this cyclic dependency between `DefaultLeaderElectionService` and the `MultipleComponentLeaderElectionDriverAdapter`. Registering a contender in the `MultipleComponentLeaderElectionService` through the adapter might result in an immediate leadership grant to the contender if the leadership was acquired beforehand. Calling `startLeaderElection` from within the constructor with `this` (the contender) as a parameter would result in the grant call being triggered on a not-fully-initialized instance.

It's essentially the same reason why we cannot move `DefaultLeaderElectionService#startLeaderElectionBackend()` into the `DefaultLeaderElectionService` constructor. We can do this after all the other refactoring is done.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/AbstractLeaderElectionService.java:
```diff
@@ -0,0 +1,86 @@
+    protected abstract boolean hasLeadership(UUID leaderSessionId);
+
+    /** {@code LeaderElectionImpl} is the default implementation of {@link LeaderElection}. */
+    private static class LeaderElectionImpl implements LeaderElection {
```
The idea of having the `LeaderElectionImpl` being implemented as an inner class was to hide the `LeaderElection`-specific interface from other implementations. The `StandaloneLeaderElectionService` would require these methods to become package-private. The `EmbeddedLeaderElectionService` is not even in the same package. But I noticed that I could achieve the same with a normal class. This improves testability later on, I guess. I created `DefaultLeaderElectionTest` with quite basic test methods for now. I hope to gradually extend it in the next (future) PRs.

zentol replied:
protected is also implicitly package-private; we wouldn't need to change anything.

XComp replied:
> and you ended up doing exactly what I asked for?...
I'm confused by the question mark here. :innocent: I found your proposal reasonable in terms of testability and went ahead with the refactoring. My [previous comment](https://github.com/apache/flink/pull/22384#discussion_r1200710974) just explained my initial reasoning. :-D

XComp commented:
I addressed the comments... PTAL

XComp commented:
Force-push rebased this PR onto #22623

XComp commented:
Force-push rebased this PR back to upstream `master` after https://github.com/apache/flink/pull/22623 was merged

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionService.java:
```diff
@@ -32,87 +29,49 @@
-    @Override
-    public synchronized boolean hasLeadership(@Nonnull UUID leaderSessionId) {
-        return hasLeadership && leaderSessionId.equals(issuedLeaderSessionId);
+    public synchronized LeaderInformation isConfirmedLeader(UUID leaderSessionID) {
```
The intention is good but I was a tad confused why isLeader and isConfirmedLeader return different types.

I feel like calling `isLeader.join()` explicitly in the test, or renaming this method to `isLeaderAndWaitForConfirm` may alleviate this.

XComp replied:
Fair enough, I guess calling `.join()` on the `CompletableFuture` is good enough. I'm gonna add JavaDoc to the `isLeader(UUID)` method, instead.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElection.java:
```diff
@@ -0,0 +1,163 @@
+     */
+    @Nullable private LeaderContender contender = null;
+
+    @Nullable private CompletableFuture<LeaderInformation> confirmationFuture = null;
```
wondering if we should cancel this in triggerContenderCleanup

XComp replied:
Ok, we need to stick around with the method still (because it's needed by `TestingLeaderElectionService.stop()`). The method will become `LeaderElection.close()` in FLINK-31785.

XComp replied:
In addition to what I mentioned in [my previous comment](https://github.com/apache/flink/pull/22384#discussion_r1202637964): This is a slight code change (comparing it to the old `TestingLeaderElectionService` where we didn't cancel the `confirmationFuture` in `TestingLeaderElectionService.stop()`. Now, we do cancel the future. But this change makes sense for the aformentioned reasons. I'm going to add a comment to document that as well

XComp commented:
FYI: the [CI failure](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=49304&view=logs&j=af184cdd-c6d8-5084-0b69-7e9c67b35f7a&t=0f3adb59-eefa-51c6-2858-3654d9e0749d&l=13641) happened due to FLINK-26635

I addressed the comments and rebased the follow-up PRs based on these changes.

XComp commented:
Thanks, @zentol . I squashed the commits and rebased the PR to the most-recent `master` for a final CI run before merging.

XComp commented:
Conflicts happened due to https://github.com/apache/flink/commit/d79f4c1d39c45ac8365fbddbe3ad7e53d42c8f41 and https://github.com/apache/flink/commit/c3409dcd0f77ef3343145ef1a05944a27f4f87ec. I rebased to the most-recent `master` and fixed the conflicts.

@reswqa just keep in mind that fixing legacy issues in leader election-related code slows down the FLINK-26522 efforts (due to possible code conflicts). Therefore, it might help to wait with things that are not really pressing issues until we're done with the PRs. Feel free to jump in and help reviewing these PRs. :-)

reswqa commented:
> Therefore, it might help to wait with things that are not really pressing issues until we're done with the PRs.

Fair enough. When I tried to modify the code to remove some legacy `ExternalResource`, I somehow forgot that it would conflict with FLINK-26522. I'm sorry to have hindered your progress and thanks for the reminder. 🙇
